[quant][core][gpu][cudnn] Added support for nhwc tensors in quantized cudnn add_relu op

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75806

Summary:
When using the the quantized cudnn add operator, if the input tensors are 4D,
cudnn requires NHWC format in v. 8.4.0 (older versions may have relaxed this constraint).
Previously, all tensors defaulted to NCHW format.

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Differential Revision: [D35651368](https://our.internmc.facebook.com/intern/diff/D35651368)

dzdang committed:
[quant][core][gpu][cudnn] Added support for nhwc tensors in quantized cudnn add_relu op

Summary:
When using the the quantized cudnn add operator, if the input tensors are 4D,
cudnn requires NHWC format in v. 8.4.0 (older versions may have relaxed this constraint).
Previously, all tensors defaulted to NCHW format.

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35651368).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -73,18 +74,28 @@ Tensor add(Tensor qa, Tensor qb, double output_scale, int64_t output_zero_point)
-      at::device(at::kCUDA).dtype(at::ScalarType::QInt8),
-      output_scale,
-      output_zero_point);
+  at::Tensor add_output = qa.dim() == 4 ? at::empty(qa.sizes(), at::device(at::kCUDA).dtype(at::kFloat), at::MemoryFormat::ChannelsLast)
+                                        : at::empty(qa.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
```
nit: there are some duplicated code here, can we have a variable to represent memoryformat and only apply the branch for the variable? like:
```
memory_format = a if xxx else b
add_output = at:empty(qa.sizes(), ..., memory_format)
```

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -73,18 +74,28 @@ Tensor add(Tensor qa, Tensor qb, double output_scale, int64_t output_zero_point)
-      output_zero_point);
+  at::Tensor add_output = qa.dim() == 4 ? at::empty(qa.sizes(), at::device(at::kCUDA).dtype(at::kFloat), at::MemoryFormat::ChannelsLast)
+                                        : at::empty(qa.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+  at::Tensor quantized_output = qa.dim() == 4 ? at::_empty_affine_quantized(qa.sizes(),
```
same here

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -73,18 +74,28 @@ Tensor add(Tensor qa, Tensor qb, double output_scale, int64_t output_zero_point)
+                                                                            output_zero_point);
   // TODO: When cudnn enables support for broadcasting, we can remove this tensor
-  at::Tensor requantize_multiplier_tensor = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+  at::Tensor requantize_multiplier_tensor = qa.dim() == 4 ? at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat), at::MemoryFormat::ChannelsLast)
```
same for all the code in this change

jerryzh168 reviewed (changes requested):
please remove duplicated code

dzdang committed:
Update on "[quant][core][gpu][cudnn] Added support for nhwc tensors in quantized cudnn add_relu op"


Summary:
When using the the quantized cudnn add operator, if the input tensors are 4D,
cudnn requires NHWC format in v. 8.4.0 (older versions may have relaxed this constraint).
Previously, all tensors defaulted to NCHW format.

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Differential Revision: [D35651368](https://our.internmc.facebook.com/intern/diff/D35651368)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35651368).

dzdang committed:
Update on "[quant][core][gpu][cudnn] Added support for nhwc tensors in quantized cudnn add_relu op"


Summary:
When using the the quantized cudnn add operator, if the input tensors are 4D,
cudnn requires NHWC format in v. 8.4.0 (older versions may have relaxed this constraint).
Previously, all tensors defaulted to NCHW format.

Test plan:
```
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Differential Revision: [D35651368](https://our.internmc.facebook.com/intern/diff/D35651368)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35651368).
